app/service/send_notification.py

```diff
@@ -172,9 +175,14 @@ def send_pdf_letter_notification(service_id, post_data):
     try:
         letter = utils_s3download(current_app.config['TRANSIENT_UPLOADED_LETTERS'], file_location)
     except S3ObjectNotFound as e:
-        current_app.logger.exception('Letter {}.pdf not in transient {} bucket'.format(
+        current_app.logger.warning('Letter {}.pdf not in transient {} bucket'.format(
             post_data['file_id'], current_app.config['TRANSIENT_UPLOADED_LETTERS'])
         )
+
+        # notification already exists e.g. if the user clicked send in different tabs
+        if get_notification_by_id(post_data['file_id']):
+            return {'id': str(post_data['file_id'])}
```

Comment:
nice. could this be up the top of the function so we get out of this fn even earlier if the notification is created and dont even check for the file?

Reply:
I thought about that but it would mean an extra unnecessary DB operation for the majority of requests. I appreciate this is just requests from the Admin app so we're not overly concerned about performance.

Do you have a strong preference either way?

Reply:
yeah, given its just admin i thought it'd be minimal. also in the happy path, it's making at least three DB queries, getting and reading a file from s3, and moving a file between buckets, so a simple primary key lookup should be realllllly quick in comparison, even if it's not necessary for 99% of the time.

it also fits in well with similar checks for, eg, if the user has permissions in admin app. You shouldn't be in this function at all since the letter is already sent - so just bail out at the top. also i just prefer exception handling clauses to be a bit more simple in general.

that said none of these views are particularly strongly held, i'll leave it up to you

Reply:
Sure, changed in https://github.com/alphagov/notifications-api/pull/3503/commits/413c6c4c2653deba6692bed74deeb0bb8c4c7f37.